lib/tools/system-commands.js

```diff
@@ -72,9 +74,62 @@ systemCallMethods.connect = async function connect (opts = {}) {
     }
   }
 
+  let port = DEFAULT_COMPANION_PORT;
+  try {
+    log.debug(`Starting companion: '${binaryPaths[IDB_COMPANION_EXECUTABLE]}'`);
+    const companionProc = new SubProcess(binaryPaths[IDB_COMPANION_EXECUTABLE], ['--udid', this.udid]);
+    companionProc.on('exit', (code, signal) => {
+      log.debug(`Companion exited with code '${code}' from signal '${signal}'`);
+    });
+
+    if (this.verbose) {
+      companionProc.on('lines-stdout', function (lines) {
+        for (let line of lines) {
+          line = line.trim();
+          if (line) {
+            log.debug(`[Companion stdout] ${line}`);
+          }
+        }
+      });
+      companionProc.on('lines-stderr', function (lines) {
+        for (let line of lines) {
+          line = line.trim();
+          if (line) {
+            log.debug(`[Companion stderr] ${line}`);
+          }
+        }
+      });
+    }
+
+    await companionProc.start(function (stdout, stderr) {
+      const out = stdout || stderr;
+
+      // check for marker that things are ready to go
+      const readyMatch = COMPANION_STARTUP_REGEXP.exec(out);
+      if (readyMatch) {
+        // find the port and save, so idb can connect
+        port = readyMatch[1];
+        return true;
+      }
+
+      // check if there has been an error
+      const errorMatch = COMPANION_STARTUP_ERROR_REGEXP.exec(out);
+      if (errorMatch) {
+        throw new Error(errorMatch[1]);
+      }
+
+      return false;
+    });
+  } catch (err) {
+    log.error(`Failed to start ${IDB_COMPANION_EXECUTABLE}: ${err.message}`);
+    throw err;
+  }
+
+  log.debug(`Companion running on port '${port}'`);
+
   try {
     try {
-      await tpExec(IDB_EXECUTABLE, ['connect', this.udid]);
+      await tpExec(IDB_EXECUTABLE, ['connect', 'localhost', port]);
```

Comment:
does it also listen on ipv6 (in case localhost is resolved to :::1)?

Reply:
I have no idea about that. If it did for connections managed by `idb` then I imagine it does this way, too.

Reply:
I mean if netstat output shows the companion is only listening ipv4 then it is better to explicitly put `127.0.0.1` as the address

Reply:
Seems like it does
```
$ netstat -an -ptcp | grep LISTEN
tcp46      0      0  *.10882                *.*                    LISTEN
...
```